### PR TITLE
Hotfix for OpenMC-OpenCG Lattice Conversion

### PR DESCRIPTION
**Pull request by wbinventor**
This is a fix for the conversion of `Lattices` to-from the OpenCG and OpenMC APIs. I thought that I had resolved this before, but apparently not. The short of it is that the *y*-dimension is indexed in opposite orders in OpenCG and OpenMC. Hence, whenever one wishes to convert an `openmc.Lattice` to an `opencg.Lattice`, or vice versa, then the `Lattice.universes` array must be reverse indexed along *y*. Although this was implemented for the `get_openmc_lattice(...)` routine, it was not implemented for `get_openmc_lattice(...)`. This PR adds a single line to `openmc.opencg_compatible.py` for the reverse indexing.

I came across this issue while trying to do data analysis on "distribcell" tallies in the following 2x2 PWR fuel assembly geometry with a water reflector:

To do this, I make use of OpenCG's region classification scheme which makes it possible to spatially locate "distribcell" tally bins. In this case, my data workflow converts the `openmc.Lattice` constructed from the "summary.h5" file into an `opencg.Lattice` just fine. However, for less obvious but necessary reasons I convert the `opencg.Lattice` back into an `openmc.Lattice` in which case the `Lattice.universes` were not appropriately indexed and the distribcell tallies could not be properly indexed.

### CODE REVIEW

**Comment by wbinventor** [comment]
Hold off on merging this for a bit, I'm going to try to make the lattice indexing more streamlined between OpenMC/CG/MOC first.

**Comment by wbinventor** [comment]
Ok, I think I discovered the *real* culprit here and it was not the OpenMC-OpenCG compatibiliy / conversion after all. Instead, the lattice distribcell offset array is dumped to the "summary.h5" file in a different order than the universes array (line 369). In particular, the universes array is reverse indexed in y, while the offsets array is not. This PR now simply reverses the y index for the offsets array in the `Summary` API.

**Comment by wbinventor** [comment]
I should add that I don't think this in issue for any of our test cases. In particular, one needs a heterogeneous lattice geometry without reflective symmetry about the y axis to break things. As far as I know, all of our test cases are relatively simple lattices with symmetry about both x and y. We might consider breaking lattice symmetries for our tests of the distribcell feature (and perhaps other features).

**Comment by paulromano** [comment]
@wbinventor Good point. We should definitely have a test with a completely asymmetric lattice. Do you want to change one of our tests as part of this PR?

**Comment by wbinventor** [comment]
Yes, it would be good to spare future developers from experiencing the same
painful debugging process if things change in the future.

What would this type of test look like? Of course it should include a
distribcell tally for an axis-asymmetric lattice (or nested lattice) with
one or more distribcells tallies - would this be a new InputSet? And I
think the results string should not only include the tally means, but also
the lattices offsets extracted from the summary file.

**Comment by paulromano** [comment]
Whether you want to do a new input set or just override the `_build_inputs` method, either is fine. And yes, extracting offsets from the summary file would be the way to go.

**Comment by wbinventor** [comment]
I just added a new test for "distribcell" tallies in an asymmetric lattice - let me know if this works for you @paulromano.

**Comment by wbinventor** [comment]
I'm not sure why my test fails on Travis for only those tests compiled with PHDF5. I've reproduced this on my local machine. When I turn off output hashing and inspect the numerical results all of the distribcell offsets look good but the distribcell tally means / std. devs. are all very different. I'm not sure why this is, I'll revive this in the morning.

**Comment by wbinventor** [comment]
I believe the test issue has been corrected. Figuring that out was again very subtle, and I'm not even sure that it didn't point to yet another bug. Basically, I was bitten since I was ignoring a few of these warnings for my new test:

```
WARNING: Maximum number of sites in fission bank reached. This can result in
          irreproducible results using different numbers of processes/threads.
```

Yes indeed, the results are irreproducible if you use parallelism with OpenMP or MPI which led to the test failure. I don't fully understand the reasons behind this warning, but I do know that it took a little tweaking of the initial source placed on this problem in order to resolve this issue.

In particular, the `lower_left` corner of the geometry is at [-32.13, -32.13, 0.] and the `upper_right` corner is at `[+32.13, +32.13, +32.13]`. However, when I specified these as the corners for my `Box` source distribution in the test I observed the warnings above (even with `only_fissionable` set to `True`). I had to shrink the box to use a lower left corner of [-32., -32., 0.] instead to resolve these warnings.

@paulromano do you have any idea if or why there might be an issue with specify an initial source distribution that extends to the very edge of the geometric boundaries? I find this particularly concerning since I use OpenCG to specify the source bounds box for most of my geometries and it extends all the way to the very edge of the geometry's bounding box.

In any case, if my discussion above does indeed point to a bug, it is separate from this PR which I believe is ready to merge.

**Comment by wbinventor** [comment]
I should also mention that the test results were *very* different regardless of the type of tally used in the test. I tried "material," "cell" and "distribcell" tallies for the original source specification and all were irreproducible. So for any other naive users out there, make sure to heed the warning about reaching the maximum number of sites in the fission bank!

**Comment by paulromano** [comment]
@wbinventor That warning usually pops up when you too few histories per thread. The fission bank is allocated to be three times larger than the source bank. When you run with multiple threads, each thread has its own source/fission bank. Let's say you have 100 histories/generation and 50 threads. Your source bank would be length 2 and your fission bank length 6. If 7 fission neutrons are created by any given thread in a generation, you'd get that warning and the 7th neutron wouldn't get stored (thus, irreproducible). If instead you had 10000 histories/generation, its much less likely that your going to store way too many neutrons because there will be a lot less variation.

So, long story short, you ought to increase the number of histories per generation. As I'm thinking about this now, perhaps we ought to just resize the fission bank when that hard limit is reached (similar to what the Vector types do).

**Comment by wbinventor** [comment]
Thanks for the explanation @paulromano, that makes a lot of sense now.

File: openmc/tallies.py [hunk 1]
```diff
@@ -2748,7 +2748,7 @@ def get_slice(self, scores=[], filters=[], filter_bins=[], nuclides=[]):
                         bin_indices.append(bin_index)
                         num_bins += 1
 
-                find_filter.bins = set(find_filter.bins[bin_indices])
+                find_filter.bins = np.unique(find_filter.bins[bin_indices])
```
Comment (paulromano) [inline openmc/tallies.py]: This function was added in NumPy 1.9. We should probably change the install_requires line in setup.py to say `numpy>=1.9`.

File: tests/test_asymmetric_lattice/test_asymmetric_lattice.py [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+        tally.add_score('nu-fission')
+
+        # Initialize the tallies file
+        tallies_file = openmc.TalliesFile()
+        tallies_file.add_tally(tally)
+
+        # Assign the tallies file to the input set
+        self._input_set.tallies = tallies_file
+
+
```
Comment (paulromano) [inline tests/test_asymmetric_lattice/test_asymmetric_lattice.py]: Delete one blank line

**Comment by paulromano** [comment]
Good to go once those two little comments are resolved

File: openmc/tallies.py (hunk 1 above)
Comment (wbinventor) [inline openmc/tallies.py]: Good point, just pushed this updated requirement.
